src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -125,22 +132,38 @@ Score(const size_t queryIndex, TreeType& referenceNode)
     const math::Range r = referenceNode.RangeDistance(queryPoint);
     minDistance = r.Lo();
     maxDistance = r.Hi();
+
+    // Check if we are a self-child.
+    if (tree::TreeTraits<TreeType>::HasSelfChildren &&
+        referenceNode.Parent() != NULL &&
+        referenceNode.Parent()->Point(0) == referenceNode.Point(0))
+    {
+      alreadyDidRefPoint0 = true;
+    }
   }
 
   const double maxKernel = kernel.Evaluate(minDistance);
   const double minKernel = kernel.Evaluate(maxDistance);
   const double bound = maxKernel - minKernel;
 
-  if (bound <= (absError + relError * minKernel) / referenceSet.n_cols)
-  {
-    // Estimate values.
-    double kernelValue;
+  // Error tolerance of the current combination of query point and reference
+  // node.
+  const double relErrorTol = relError * minKernel;
+  const double errorTolerance = absErrorTol + relErrorTol;
+
+  // We relax the bound for pruning by accumError(queryIndex), so that if there
+  // is any leftover error tolerance from the rest of the traversal, we can use
+  // it here to prune more.
+  double pointAccumErrorTol;
+  if (alreadyDidRefPoint0)
+    pointAccumErrorTol = accumError(queryIndex) / (refNumDesc - 1);
+  else
+    pointAccumErrorTol = accumError(queryIndex) / refNumDesc;
```

Comment:
If you wanted, you could rewrite this like this:

```
const double pointAccumErrorTol = accumError(queryIndex) /
    ((alreadyDidRefPoint0) ? (refNumDesc - 1) : refNumDesc);
```

It's up to you which you think is more readable. :+1:

Reply:
To be honest I'm not a big fan of the ternary operator, I feel like it's less readable :sweat_smile:.